get_openstack_info_open_port/get_openstack_info_open_port/get_openstack_info_open_port.py

```diff
@@ -0,0 +1,190 @@
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.settimeout(10)
+        result = sock.connect_ex((self.ip, self.port))
+        sock.close()
+        if result == 0:
+            self._log.debug(f"Port '{self.port}' is reachable.")
+            return True
+        else:
+            self._log.debug(f"Port '{self.port}' is NOT reachable.")
+            sys.exit(3)
+       
+    def _get_credentials(self):
+        """
+        Load login information from environment
+        :returns: credentials
+        :rtype: dict
+        """
+        cred = dict()
+        cred['auth_url'] = os.environ.get('OS_AUTH_URL').replace("v2.0", "v3")
+        cred['username'] = os.environ.get('OS_USERNAME')
+        cred['password'] = os.environ.get('OS_PASSWORD')
+        if 'OS_PROJECT_ID' in os.environ:
+            cred['project_id'] = os.environ.get('OS_PROJECT_ID')
+        if 'OS_TENANT_ID' in os.environ:
+            cred['project_id'] = os.environ.get('OS_TENANT_ID')
+        cred['user_domain_name'] = os.environ.get(
+            'OS_USER_DOMAIN_NAME', 'default')
+        for key in cred:
+            if not cred[key]:
+                self.logger.critical(
+                    'Credentials not loaded to environment: did you load the rc file?')
+                exit(1)
+        return cred
+
+    def _os_auth(self):
+        self.keystone_session = session.Session(
+            auth=v3.Password(**self._get_credentials()))
+        self.keystone_v3 = keystoneclient_v3.Client(
+            session=self.keystone_session)
+        self.nova = nova.Client("2.1", session=self.keystone_session)
+        self.neutron = neutron.Client("2.0", session=self.keystone_session)
+        
+    def _check_openstack(self):
+        ''' Check that OpenStack variables are loaded '''
+        if 'OS_USERNAME' not in os.environ:
+            self._log.error("You don't seem to have loaded a source file with credentials to OpenStack. Please do so, and try again.")
+            sys.exit(1)
+
+
+    def _get_openstack_instances(self):
+        self.server_list = self.nova.servers.list(detailed=True, search_opts={"all_tenants": True})
+        self._log.debug(f"Obtained {len(self.server_list)} servers.")
+        
+        
+    def _locate_affected_instance(self):
+        self.affected_instance = None
+        for server in self.server_list:
+            server_info = server.to_dict()
+            for project in server_info['addresses'].keys():
+                for address in server_info['addresses'][project]:
+                    if address['addr'] == self.ip:
+                        self._log.debug(f"Located instance '{server_info['name']}' with ID '{server_info['id']}' and IP '{address['addr']}' in project '{project}'")
+                        self.affected_instance = server_info
+                        #print(json.dumps(self.affected_instance, indent=2))
```

Comment:
You can do a floating ip show -> port show and get the device id ( server id)
This might be  nicer as it needs to pull only a single item of data each time.

Reply:
Added an alternative method to test the speed of both. I have to get all floating IPs, because the API call to get a floating IP needs the ID of the floating IP and can't search by IP address. I can't test in production.